### ISSUE TITLE
iam_role introduced in v0.17.2 does not have behave like --terragrunt-iam-role cli flag

### ISSUE THREAD

**Issue by conzy** [gruntwork-io/terragrunt#667]
The documentation says these are equivalent but they are not. It seems like when using the `--terragrunt-iam-role` cli flag or exporting the `TERRAGRUNT_IAM_ROLE` env var the role assumption happens early in the terragrunt flow and sets up my remote state backend correctly. This is the expected behaviour

`aws-vault exec bastion -- terragrunt plan --terragrunt-iam-role arn:aws:iam::1111111111:role/admin`

In this case the bastion account ID is 00000000 and the target account ID is 1111111111

This all works as expected

However when I set the iam_role in the terragrunt.tfvars file like this:

```
terragrunt = {
  iam_role = "arn:aws:iam::1111111111:role/admin"
  terraform {
    source = "git@github.com:foo/foo-modules.git//terragrunt_iam_bastion?ref=v0.8.26"
  }

  # Include all settings from the root terraform.tfvars file
  include = {
    path = "${find_in_parent_folders()}"
  }
}

aws_region    = "eu-west-1"
account_alias = "foo-bastion"
```

The role assumption seems to happen later in the terragrunt flow. SO terragrunt thinks I am configuring a new remote state backend. e.g when calling

`aws-vault exec bastion -- terragrunt plan`

It wants to reconfigure the remote state backend to be that of the bastion account

```
Backend config has changed from 
map[bucket:foo-11111111111-state dynamodb_table:terraform-locks encrypt:%!s(bool=true) key:organizations/bastion/global/iam/terraform.tfstate region:eu-west-1] to 
map[region:eu-west-1 dynamodb_table:terraform-locks encrypt:%!s(bool=true) bucket:foo-0000000000-state key:organizations/bastion/global/iam/terraform.tfstate]
```

The remote state backend configuration in my root terraform.tfvars looks like this:

```
terragrunt = {
  remote_state {
    backend = "s3"
    config {
      encrypt        = true
      bucket         = "foo-${get_aws_account_id()}-state"
      key               = "${path_relative_to_include()}/terraform.tfstate"
      region         = "eu-west-1"
      dynamodb_table = "terraform-locks"
    }
  }
```

Let me know if you need any additional info

**Comment by conzy** [comment]
Just wondering is there anything I can do to add additional context around this issue? I know lots of people use `aws-vault` with terragrunt and I'd love to be able to use the `iam_role` configuration in my terraform.tfvars file for added safety / convenience

**Comment by brikis98** [comment]
Sorry, we've been absolutely buried and haven't had time to dig into this. If you have suggestions on how to fix this, a PR is very welcome!

**Comment by StevePavlin** [comment]
Ran into this today as well, currently using Docker for different environments to bake in the assume role complexity. Would love to get rid of the containers and put the ARN in a terragrunt.hcl file for each environment.

**Comment by sebastianmacarescu** [comment]
Anything new on this?

**Comment by brikis98** [comment]
@denis256 could you look into this?

**Comment by denis256** [comment]
Found that  `get_aws_account_id()` returns IAM details defined by environment variables/terragrunt arguments, ignoring value from `iam_role`

```
$ aws sts get-caller-identity
{
    "UserId": "AAA",
    "Account": "BBB",
    "Arn": "arn:aws:iam::YYYY:/app0"
}

$ cat terragrunt.hcl
  iam_role = "arn:aws:iam::XXXX:role/app1"

  remote_state = {
    backend = "s3"
    config = {
      encrypt = true
      bucket = "foo-${get_aws_account_id()}-state"
      key = "${path_relative_to_include()}/terraform.tfstate"
      region = "us-east-1"
      dynamodb_table = "terraform-locks"
    }
  }

$ terragrunt init 
Remote state S3 bucket foo-YYY-state does not exist or you don't have permissions to access it. Would you like Terragrunt to create it? (y/n) 
```

I think the expectation is to create `foo-XXXX-state` S3 bucket

**Comment by brikis98** [comment]
Yup, I think that would be the bug, right?

**Comment by denis256** [comment]
Yes, the issue is that `${get_aws_account_id()}` and any other expression get evaluated before the entire file is processed, so when the value of `get_aws_account_id()` is calculated `iam_role` wasn't parsed and loaded to evaluation context...

So far I will try to explore the direction of changing how the file is processed, to have for each `${...}` expression "promises" which will be resolved after initial file evaluation...

**Comment by brikis98** [comment]
Ah, right, that makes sense. We already have some order dependence in Terragrunt parsing... But I guess `get_aws_account_id` could appear anywhere in the code, so it's tricky to know when to process it. 

Taking a step back, I wonder if adding `get_aws_account_id` was a mistake in general. Same for `get_aws_caller_identity_arn` and `get_aws_caller_identity_user_id`, which likely exhibit the same bug. All of these functions are very AWS specific. Doing AWS-specific things makes sense in the remote state logic, as S3 is one of the places remote state is supported, but as a general feature in Terragrunt, it seems odd to have it. 

Perhaps what we really need, as discussed before, is a first-class way to use data sources with Terragrunt. That way, you could use, for example, the [`aws_caller_identity` data source](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/caller_identity), [`aws_vpc` data source](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/vpc), or any other data source for any cloud... 

@yorinasub17 and @infraredgirl curious to hear your thoughts.

**Comment by yorinasub17** [comment]
> So far I will try to explore the direction of changing how the file is processed, to have for each ${...} expression "promises" which will be resolved after initial file evaluation...

I believe the functions handle the IAM role correctly if it is assigned in the `terragruntOptions` struct (the `IamRole` attr), so it might be simpler to just parse out the `iam_role` config first in the `DecodeBaseBlocks` function and then assign that attribute on `terragruntOptions` prior to parsing the rest of the config. That might be simpler approach than introducing lazy evaluation.

> Perhaps what we really need, as discussed before, is a first-class way to use data sources with Terragrunt. That way, you could use, for example, the aws_caller_identity data source, aws_vpc data source, or any other data source for any provider?

I think the data sources are likely to have the same problem here, as the real issue is about authenticating terragrunt to the cloud during config parsing. That is, the functions (and data sources for that matter) are things that need to be authenticated during the parsing stage, but config options for authenticating to the cloud like `iam_role` is only handled after parsing. So supporting data sources feels like an orthogonal discussion to fixing this bug.
